### DISCUSSION TITLE
Make Twirl render custom scala types

### DISCUSSION THREAD

**Discussion by hertg** [playframework/playframework#12482] [Q&A]
Is it possible to customize the rendering of a Scala class in Twirl, other than Twirl simply calling the class' `toString` method?
https://www.playframework.com/documentation/3.0.x/ScalaTemplates#Displaying-Scala-types

---

_Example_

```scala
import play.twirl.api.Html

// this is a custom type i created
trait HtmlElement {
  def render: Html
}
```

```scala
import play.twirl.api.Html

// this is a custom class i created
case class Btn(text: String) extends HtmlElement {
  override def render: Html = Html(s"<button>$text</button>")
}
```

When a class of type `HtmlElement` is used in Twirl, I'd like it to be calling the `render` method and include the return value unescaped, instead of doing a `toString`.

```
@Btn("foo")
// this renders: Btn(foo)
```

```twirl
@Btn("foo").render
// this renders <button>foo</button>
```

I'd like those two calls to both result in the HTML output.

**Comment by mkurz** [comment]
> I'd like those two calls to both result in the HTML output.

I don't think this is easily possible with html format. If you look at the twirl scala code that gets generated in your target folder and debug what is getting called, you will end up [here](https://github.com/playframework/twirl/blob/2.0.4/api/shared/src/main/scala-2.13/play/twirl/api/BaseScalaTemplate.scala#L40):

```scala
      case v if v != null => format.escape(v.toString)
```

so if twirl does not know about the special class you are using, it will always call it's `toString` method and then it will escape that string returned. So even if you do
```scala
trait HtmlElement {
  def render: Html
  override def toString: String = render.toString // will give you raw html
}
```
the `format.escape` call later will then escape that raw html. This is also not what you want.

Twirl's HTML format only outputs raw html when you pass a `Html` object ([this line](https://github.com/playframework/twirl/blob/2.0.4/api/shared/src/main/scala-2.13/play/twirl/api/BaseScalaTemplate.scala#L23)). so that means either make `Btn` a subclass of `Html` or there is no way around of calling `render` or any other method that gives a an `Html` instance.
To be honest, I think it is a good thing to explicitly call a method to output raw html, because usually it's much safer if html output is unescaped per default.


_Maybe_ there is a way to automatically convert your `Btn` or `HtmlElement` with some implicit magic, but not 100% sure, I didn't really try a lot.
You could e.g. try to set up an object `TemplateImplicits` and import it in `build.sbt` with 
```sbt
TwirlKeys.templateImports += "views.html.TemplateImplicits._"
```
and here I have an example
```scala
package views.html

import play.twirl.api.Html
import play.twirl.api.HtmlFormat

object TemplateImplicits {
  // does not work
  implicit def htmlElement2html(he: HtmlElement): Html = he.render

  // this does work so you can call @Btn("foo").raw but does not give you a benefit over calling render...
  implicit class RichHtmlElement(he: HtmlElement) {
    def raw(): Html              = he.render
  }
}
```
